### PR TITLE
New Theme for Docs

### PR DESCRIPTION
**Pull request by ahill187**
#### GitHub

- [x] I've given this PR a concise, self-descriptive, and meaningful title
- [ ] I've linked relevant issues in the PR body
- [ ] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [ ] I've applied a [release milestone](https://github.com/neuropoly/spinalcordtoolbox/milestones) (major, minor, patch) in line with [Semantic Versioning guidelines](https://github.com/neuropoly/spinalcordtoolbox/wiki/Misc%3A-Creating-a-new-release#convention-for-naming-releases) 
- [x] I've assigned a reviewer

#### PR contents

- [ ] I've consulted [SCT's internal developer documentation](https://github.com/neuropoly/spinalcordtoolbox/wiki) to ensure my contribution is in line with any relevant design decisions
- [ ] I've added [relevant tests](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/neuropoly/spinalcordtoolbox/wiki/Programming%3A-Documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

## Description
1. This pull request changes the style for the `ReadtheDocs` documentation to the `Furo` theme: https://pradyunsg.me/furo/quickstart/
2. I've also rearranged the installation instructions to make them a bit easier to follow.
3. There is now a copy option for all code chunks (little button in the top right)

Feedback welcome on the style!

### CODE REVIEW

**Comment by joshuacwnewton** [comment]
Out of curiosity, what was the motivation for this change? Is there a specific issue with the existing theme that the new theme fixes?

File: documentation/source/overview/references.rst [hunk 1]
```diff
@@ -117,95 +111,193 @@ Applications
...
--  `Tinnermann et al. Observation of others’ painful heat stimulation involves responses in the spinal cord. Science Advances, 2021 <https://pubmed.ncbi.nlm.nih.gov/33789899/>`_
--  `Zhang et al. Automatic spinal cord segmentation from axial-view MRI slices using CNN with grayscale regularized active contour propagation. Computers in Biology and Medicine, 2021 <https://pubmed.ncbi.nlm.nih.gov/33780869/>`_
--  `Savini et al. Pilot Study on Quantitative Cervical Cord and Muscular MRI in Spinal Muscular Atrophy: Promising Biomarkers of Disease Evolution and Treatment? Front Neurol, 2021 <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8039452/>`_
\ No newline at end of file
+.. list-table::
+   :widths: 80 20
+   :header-rows: 1
+
+   * - Citation
+     - Year
```
Comment (joshuacwnewton) [inline documentation/source/overview/references.rst:119]: Some small comments on the table:

* #3349 separates this section off into its own page (`studies.rst`). So, if changes are made in `references.rst`, they'll conflict.
* The year column is a bit wider than it maybe needs to be.
Screenshot from 2021-04-24 16-21-56
* Right now, the year info is duplicated, I think.
* What do you think of having the columns be Author, Title, Journal/Conference, and Year?

(Apologies if this is a WIP and my comments are premature here. :sweat_smile:)

**Review by jcohenadad — COMMENTED** [review]
Thank you for this initiative @ahill187 !

A few comments:

Pros:
- I *LOVE* the added sub-TOC that appears on the right. It makes it much easier to follow the page.
- I *LOVE* the new "copy" button on the code blocks
- (unrelated to the theme): I really like your improvement on the installation page. It is waaaay clearer.

Cons:
- I agree with @joshuacwnewton the dark mode (which I personally never use) doesn't look great.
- (minor) I find the new fonts make it more difficult to distinguish the headings 1, 2, etc. (conversely, the RTD theme has very clear headings). 
- (minor) I personally quite like the current RTD fonts. Moreover, given that the default RTD theme is widely used, I feel that many people are used to the style and thus feel "at home" when they encounter a new website. When I first discovered this new SCT webpage cosmetic, it took me some time to adjust to stupid things like: where are the admonitions, what are subheadings 1, 2, etc., the code block with black background is surprising (and thus distracted me from the actual content of the page).
- installation for macos: i would not start the requirements by saying that ppl need to install gcc, because in most cases it is _already_ installed. Instead I would maybe just ask for people to check if it is indeed installed. Also, i think we had some issues about that (no time to dig), but one of the conclusion was that gcc is most of the time not required.

File: documentation/source/overview/references.rst (hunk 1 above)
Comment (joshuacwnewton) [reply documentation/source/overview/references.rst:119]: Update: The table idea was nixed (due to wanting to adhere to academic-style reference formatting), but some alternate formatting ideas were proposed in https://github.com/neuropoly/spinalcordtoolbox/issues/3357

**Comment by ahill187** [comment]
> Out of curiosity, what was the motivation for this change? Is there a specific issue with the existing theme that the new theme fixes?
> 
> As for feedback, there are some parts of the theme I really like, namely the built-in table of contents for each page in the top right (awesome, and very useful).
> 
> There are also some parts that I don't like as much (subjective opinions of course, though!). The dark theme is a bit harsh to my eyes, especially because of the blue accents. Also, some of our graphics either have not enough contrast (the transparent SCT logo) or too much contrast (any graphics with a white background).
> 
> Screenshot from 2021-04-24 12-32-10
> 
> Lastly, the theme (as it is now) has a narrow content width, which clashes with some of our layouts. (Previous discussion: [#3349 (comment)](https://github.com/neuropoly/spinalcordtoolbox/pull/3349#discussion_r618650420))

The motivation was just trying something new out! I'm not sure what you mean about the dark theme, on my page it is a light theme?

sct_theme

**Comment by ahill187** [comment]
> Thank you for this initiative @ahill187 !
> 
> A few comments:
> 
> Pros:
> 
> * I _LOVE_ the added sub-TOC that appears on the right. It makes it much easier to follow the page.
> * I _LOVE_ the new "copy" button on the code blocks
> * (unrelated to the theme): I really like your improvement on the installation page. It is waaaay clearer.
> 
> Cons:
> 
> * I agree with @joshuacwnewton the dark mode (which I personally never use) doesn't look great.
> * (minor) I find the new fonts make it more difficult to distinguish the headings 1, 2, etc. (conversely, the RTD theme has very clear headings).
> * (minor) I personally quite like the current RTD fonts. Moreover, given that the default RTD theme is widely used, I feel that many people are used to the style and thus feel "at home" when they encounter a new website. When I first discovered this new SCT webpage cosmetic, it took me some time to adjust to stupid things like: where are the admonitions, what are subheadings 1, 2, etc., the code block with black background is surprising (and thus distracted me from the actual content of the page).
> * installation for macos: i would not start the requirements by saying that ppl need to install gcc, because in most cases it is _already_ installed. Instead I would maybe just ask for people to check if it is indeed installed. Also, i think we had some issues about that (no time to dig), but one of the conclusion was that gcc is most of the time not required.

Dark theme: Ah good point, I didn't realize there was a dark theme option! I agree, that doesn't look too great.
Headings: Yes actually that was something I wanted to fix too, they're not quite clear.
gcc: Got it, I'll change that!
Background/Fonts: I was modelling the dark code background off GitBook, but that's always something we can discuss/change. Same with the fonts :)

**Comment by joshuacwnewton** [comment]
Regarding 6b1a874, I love this change! Here is some previous discussion for a similar idea: https://github.com/neuropoly/spinalcordtoolbox/issues/3095.

**Comment by ahill187** [comment]
@joshuacwnewton Regarding the discussion of width, I think the maximum recommended width is partly due to screen sizes (which Julien pointed out are generally large), but also due to optimizing reading comprehension/readability. 

> If line lengths are too long, the return sweeps to the beginning of the next line are difficult. If the lines are too short, readers cannot make use of much information in each fixation

Citation: https://www.sciencedirect.com/science/article/abs/pii/S1071581901904586

Study with school-aged children: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3777945/

Short version: https://www.smashingmagazine.com/2014/09/balancing-line-length-font-size-responsive-web-design/

Basically, it seems like there may be a balance to be struck for optimizing readability, where lines which are too long or too short can reduce readability.

**Comment by ahill187** [comment]
> Regarding [6b1a874](https://github.com/neuropoly/spinalcordtoolbox/commit/6b1a8747b70c8bb6aca55408ff640e631c3e1320), I love this change! Here is some previous discussion for a similar idea: #3095.

Oh awesome, I didn't even see that issue! I am less familiar with the SCT code so if you have any suggestions about how to split it up differently/more optimally, let me know. I basically just took every heading and moved it into its own file.

File: documentation/source/overview/references.rst (hunk 1 above)
Comment (ahill187) [reply documentation/source/overview/references.rst]: Ah okay, I removed the table, and perhaps we can do reformatting in a separate pull request?

**Comment by jcohenadad** [comment]
> Do you prefer the code blocks from this page: https://pradyunsg.me/furo/quickstart/?

Yes 😊

**Comment by jcohenadad** [comment]
just in case you didn't notice:
image

**Comment by jcohenadad** [comment]
> Citation: https://www.sciencedirect.com/science/article/abs/pii/S1071581901904586
Study with school-aged children: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3777945/
Short version: https://www.smashingmagazine.com/2014/09/balancing-line-length-font-size-responsive-web-design/
Basically, it seems like there may be a balance to be struck for optimizing readability, where lines which are too long or too short can reduce readability.

Thank you so much for this in-depth investigation @ahill187 ! This is super interesting, and yes, it is convincing. I am definitely sold on the idea to reinstate the width limitation (maybe less constraining than the previous one though)

**Comment by ahill187** [comment]
> > Do you prefer the code blocks from this page: https://pradyunsg.me/furo/quickstart/?
> 
> Yes 😊

Okay, I can change the colour!

**Comment by ahill187** [comment]
> just in case you didn't notice:
> image

Oh good catch, thank you I hadn't noticed that!

**Comment by jcohenadad** [comment]
@ahill187 i was looking again at the updated RTD, and I find that the new code block is not super-well distinguishable from the white background:
image

Do you agree? I am wondering if we could have a code block with a slightly darker background maybe?

**Comment by ahill187** [comment]
> @ahill187 i was looking again at the updated RTD, and I find that the new code block is not super-well distinguishable from the white background:
> image
> 
> Do you agree? I am wondering if we could have a code block with a slightly darker background maybe?

Yah you're right, it is a tad light. I can make it a bit darker 😃

**Comment by joshuacwnewton** [comment]
Ah! There are a few different web utilities that allow you to check contrast, because this is also an [a11y](https://www.a11yproject.com/) design issue. 

For example: https://webaim.org/resources/contrastchecker/

**Comment by jcohenadad** [comment]
I prefer the code box on [this website](https://torchio.readthedocs.io/datasets.html)

**Comment by joshuacwnewton** [comment]
One more note about colour contrast: On Firefox 88.0, Ubuntu 20.04.2 LTS, I see a dark theme by default and the following white-on-white for code blocks and admonitions:

<details>

Screenshot from 2021-05-06 20-24-16
Screenshot from 2021-05-06 20-24-25
</details>

On Chrome, though, things are OK. I see a light theme by default (and it looks quite nice!)

<details>

Screenshot from 2021-05-06 20-25-30
</details>

**Comment by joshuacwnewton** [comment]
I've linked #3095 since this PR currently contains very similar work (splitting up into subpages).

**Comment by ahill187** [comment]
> additional comments:
> 
> * I prefer the code box on [this website](https://torchio.readthedocs.io/datasets.html)
> * Can we try to match the RTD fonts?
> * can you put the same background color as in the RTD
> * we can try a dark background on the TOC

Changed background colour to RTD here: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/852e2fd76f67c104003fbec6589f4a80561d1073

**Comment by valosekj** [comment]
Hi, I have tried this [new theme](https://spinalcordtoolbox--3355.org.readthedocs.build/en/3355/) and I really like it! 👍🏻  It has great freshly looking design. I also really like the navigation panel on the right side of the page. For instance, the [CLI tools](https://spinalcordtoolbox--3355.org.readthedocs.build/en/3355/user_section/command-line.html#command-line-tools) section is more clear with this side panel compare to the old theme.
I tried the new theme also on my iPhone and it looks good as well (which is important because I sometimes browse documentations on my mobile device).

Only thing; when I open the new theme in Safari or Chrome on my Mac, I have the main text’s color set to white (same as background) and I have to select the text to see it, see screenshot below. But I guess that this will be fixed.
Maybe, I personally prefer websites with complete dark-mode, i.e., I would switch the central part of the website to black background. Completely dark-mode websites are more pleasant for reading during nights. But it’s only my personal preference.

One minor thing; SCT logo in the left up corner looks less contrast when is placed on the dark background.

image

**Comment by jcohenadad** [comment]
Thank you for your feedback @valosekj 🙏

**Comment by jcohenadad** [comment]
About the darkmode, the [torchio website](https://torchio.readthedocs.io/quickstart.html), which uses similar template, manages it well:

image

**Comment by rosella1234** [comment]
Hi all,
by inspecting the [new theme](https://spinalcordtoolbox--3355.org.readthedocs.build/en/3355/overview/introduction.html), I found it really attractive and easy-to-read. 
Just one thing: I think contrast between titles and explanations below is a bit too low: I would slightly thicken titles (see screenshot below)
image

**Comment by jcohenadad** [comment]
Thank you for your feedback @rosella1234 🙏

**Comment by ahill187** [comment]
> About the darkmode, the [torchio website](https://torchio.readthedocs.io/quickstart.html), which uses similar template, manages it well:
> 
> image

I'll take a look at implementing a dark mode like this 😄

**Comment by jcohenadad** [comment]
https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/57bea12df5b0eafcd573b83e1df1d9c8fad34252 i like the orange 😊

**Comment by ahill187** [comment]
So the dark theme option (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/d50e2480f36195a0e9945a9dd87ee43d027c5295) should be good now, let me know if you have any feedback @jcohenadad @joshuacwnewton @valosekj ! 😃

**Comment by jcohenadad** [comment]
> So the dark theme option ([d50e248](https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/d50e2480f36195a0e9945a9dd87ee43d027c5295)) should be good now, let me know if you have any feedback @jcohenadad @joshuacwnewton @valosekj ! 😃

Indeed it is much better 🎉  

I just found a minor glitch:
image

**Comment by ahill187** [comment]
> > So the dark theme option ([d50e248](https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/d50e2480f36195a0e9945a9dd87ee43d027c5295)) should be good now, let me know if you have any feedback @jcohenadad @joshuacwnewton @valosekj ! 😃
> 
> Indeed it is much better 🎉
> 
> I just found a minor glitch:
> image

Oh yes! Good catch, thanks!

**Comment by jcohenadad** [comment]
I think we are close to a final version!  I just have two very minor comments:
- Would it be possible to have the section on the TOC (OVERVIEW, USER SECTION, etc.) in bold?
- Would it be possible to have admonition with the same font size? I find it distracting that the font is smaller.
- We should not forget to also update the logo

File: setup.py [hunk 2]
```diff
@@ -42,9 +42,11 @@
     include_package_data=True,
     extras_require={
         'docs': [
-            'sphinx',
+            'sphinx==4.0.0',
```
Comment (joshuacwnewton) [inline setup.py:45]: We should keep an eye on this before merging! :slightly_smiling_face:

File: setup.py [hunk 3]
```diff
@@ -42,9 +42,11 @@
     include_package_data=True,
```
Comment (joshuacwnewton) [inline setup.py:42]: The contrast is low for the code-formatted text in the light theme, which makes it a little hard to read, especially if it's used in a title (and thus shows up on the sidebar)

Screenshot from 2021-07-16 12-22-52

Compare to dark theme:

Screenshot from 2021-07-16 12-24-34

Comment (joshuacwnewton) [inline setup.py:42]: The hover-over text for copying code snippets overlaps with transparency if the code snippet is too long:

Screenshot from 2021-07-16 12-19-47

Could we have an opaque background for the "Copy" text?

Comment (joshuacwnewton) [inline setup.py:42]: Toctree captions are small (relative to the test), and in all caps. I think this is also how section headers used to be, too, but they were tweaked... can we also tweak toctree captions?

```rst
# Code to generate toctree with caption
.. toctree::
   :caption: Table of Contents
```

Screenshot from 2021-07-16 12-27-57

**Review by joshuacwnewton — COMMENTED** [review]
I just have some very minor nitpicks I came across while looking through. But, overall it looks very good! :slightly_smiling_face:

File: setup.py (hunk 2 above)
Comment (ahill187) [reply setup.py:45]: Yes good flag, this was a temporary fix but yah we should check if their bug is better now 😄

File: setup.py (hunk 3 above)
Comment (ahill187) [reply setup.py:42]: Agreed, that looks much better in the dark theme!

Comment (ahill187) [reply setup.py:42]: I think so! I'll investigate 🔎

**Comment by ahill187** [comment]
> I think we are close to a final version! I just have two very minor comments:
> 
> * Would it be possible to have the section on the TOC (OVERVIEW, USER SECTION, etc.) in bold?
> * Would it be possible to have admonition with the same font size? I find it distracting that the font is smaller.
> * We should not forget to also update the logo

1. Yup we can make the section TOC bold 👍 
2. Yes that bothered me too, I can change the admonition font
3. Yes need to do the logo. I was going to ask about that, how was the logo generated/how can we edit it properly?

File: setup.py (hunk 3 above)
Comment (ahill187) [reply setup.py:42]: Changed here: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/452a37076055c0c931c163b967a2dafc4618679b

Comment (joshuacwnewton) [reply setup.py:42]: Perfect! :tada:

**Comment by jcohenadad** [comment]
> So the TOC headers are already bold, I could make them slightly larger?

yes! it's better

> Yes that bothered me too, I can change the admonition font

I also find it better.

About the admonition, I am wondering why this one has a proper warning sign:

image

But this one has not:

image

Maybe the admonition just needs to be switched for "important"? 

> Yes need to do the logo. I was going to ask about that, how was the logo generated/how can we edit it properly?

It was generated by Keynote. All logos/figures are [here](https://drive.google.com/open?id=1LI9rJf-7mikvaVuOMoCyc3wpBJFc9PMV) (only accessible by the SCT team which you should be part of @ahill187 ). 

Minor additional comment:

- I would change [Support](https://spinalcordtoolbox--3355.org.readthedocs.build/en/3355/user_section/support.html) for "Help" because support reminds me of financial support/sponsoring, etc.

File: documentation/source/_static/css/custom.css [hunk 4]
```diff
@@ -9,5 +17,286 @@
 .wy-table-responsive {overflow: visible !important;}
 .wy-table-responsive table td, .wy-table-responsive table th  {white-space: normal; !important;}
 
+.tutorials-header .header-logo {
+  background-image: url("../../../../imgs/logo_sct.png");
+}
+
+body {
+  /* font-family: FreightSans, Helvetica Neue, Helvetica, Arial, sans-serif; */
+  font-family: 'Lato', sans-serif, "proxima-nova","Helvetica Neue",Arial;
```
Comment (jcohenadad) [inline documentation/source/_static/css/custom.css:26]:
```suggestion
  font-family: 'Lato', sans-serif, "proxima-nova", "Helvetica Neue", Arial;
```

File: documentation/source/_static/css/custom.css [hunk 5]
```diff
@@ -9,5 +17,286 @@
...
+    }
+    .highlight .nv {
+        color: #40ffff;
+    }
+    .highlight .s1 {
+        color: #ed9d13;
+    }
+}
+
+
```
Comment (jcohenadad) [inline documentation/source/_static/css/custom.css:61]:
```suggestion
```

File: documentation/source/_static/css/custom.css [hunk 6]
```diff
@@ -9,5 +17,286 @@
 .wy-table-responsive {overflow: visible !important;}
 .wy-table-responsive table td, .wy-table-responsive table th  {white-space: normal; !important;}
 
+.tutorials-header .header-logo {
+  background-image: url("../../../../imgs/logo_sct.png");
```
Comment (jcohenadad) [inline documentation/source/_static/css/custom.css:21]: should this also move to `source/_static/img`?

**Comment by joshuacwnewton** [comment]
> About the admonition, I am wondering why this one has a proper warning sign:

I think I know what happened here. A while back, we had experimented with removing the "!" symbol from admonitions because of the Testimonials page. (On the old theme, even "Note" had a "!" symbol.) So, to compensate, a warning emoji was manually added to the Medical Disclaimer admonition.

But, I agree that now the admonition type can be changed.

**Comment by jcohenadad** [comment]
> > About the admonition, I am wondering why this one has a proper warning sign:
> 
> I think I know what happened here. A while back, we had experimented with removing the "!" symbol from admonitions because of the Testimonials page. (On the old theme, even "Note" had a "!" symbol.) So, to compensate, a warning emoji was manually added to the Medical Disclaimer admonition.
> 
> But, I agree that now the admonition type can be changed.

done in 84b41c72abf397070410acaaf0d9b14bbee77451

**Comment by jcohenadad** [comment]
@ahill187 is there anything else that needs to be addressed here? maybe it is a good time to merge?

**Comment by joshuacwnewton** [comment]
> is there anything else that needs to be addressed here?

Just a quick note: The failing tests are due to an issue that was fixed in master, so I'm just going to update the branch so no one spends time looking into that. :sweat_smile:

File: setup.py (hunk 2 above)
Comment (ahill187) [reply setup.py]: Removed the pin to `sphinx==4.0.0` here: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/1d3e5ac921bad27bc8f68906cf24bb951822bd22

File: documentation/source/_static/css/custom.css (hunk 6 above)
Comment (ahill187) [reply documentation/source/_static/css/custom.css]: Done: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/b4666cd317a8496f8663c7c6f07e1d763e8c0129

**Comment by ahill187** [comment]
> > So the TOC headers are already bold, I could make them slightly larger?
> 
> yes! it's better
> 
> > Yes that bothered me too, I can change the admonition font
> 
> I also find it better.
> 
> About the admonition, I am wondering why this one has a proper warning sign:
> 
> image
> 
> But this one has not:
> 
> image
> 
> Maybe the admonition just needs to be switched for "important"?
> 
> > Yes need to do the logo. I was going to ask about that, how was the logo generated/how can we edit it properly?
> 
> It was generated by Keynote. All logos/figures are [here](https://drive.google.com/open?id=1LI9rJf-7mikvaVuOMoCyc3wpBJFc9PMV) (only accessible by the SCT team which you should be part of @ahill187 ).
> 
> I've regenerated it using background color #3d3d3c (got it from [here](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/452a37076055c0c931c163b967a2dafc4618679b/documentation/source/_static/css/custom.css#L276)) in [e00aefe](https://github.com/spinalcordtoolbox/spinalcordtoolbox/commit/e00aefe443cebf2729c5a6f8d7ade183aa17ae33).
> 
> Minor additional comment:
> 
> * I would change [Support](https://spinalcordtoolbox--3355.org.readthedocs.build/en/3355/user_section/support.html) for "Help" because support reminds me of financial support/sponsoring, etc.

Great! Thanks for doing the logo 😄 I renamed 'support' to 'help', I agree that it makes more sense: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/4051a670bcd8e2773de16327aa5adee34fa8ac6b

File: setup.py (hunk 3 above)
Comment (ahill187) [reply setup.py]: Changed here: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/3824ece8231dafcb7e21a98e81e7eabe001ab5c0

Comment (ahill187) [reply setup.py]: Thanks, that's a great point! I changed it here: https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3355/commits/665784319432c3a29c692ce498b1534c57b515a4

**Comment by ahill187** [comment]
> @ahill187 is there anything else that needs to be addressed here? maybe it is a good time to merge?

I think we've addressed everything? Seems ready to merge!

**Review by jcohenadad — APPROVED** [review]
thank you @ahill187 for the amazing contribution!!! 🙏
